### DISCUSSION TITLE
How does it not re-render <TextInput> in react-native without direct manipulation?

### DISCUSSION THREAD

**Discussion by Noitidart** [react-hook-form/react-hook-form#5111] [General] [answered]
In the example on the homepage for react-native we see the way to use this lib in react-native is the code below. react-hook-form doesn't seem to be attaching a ref and doing direct manipulation https://reactnative.dev/docs/direct-manipulation#setnativeprops-to-clear-textinput-value (ie: `textInputRef.current.setNativeProps({ text: '' })`). So how is react-hook-form changing the value without causing re-renders?

```
import { TextInput } from 'react-native';

function Input({ control, name ) {
  const { field } = useController({ control, name, defaultValue: '' });

  return (
    <TextInput value={field.value} onChangeText={field.onChange} />
  )
}

function App() {
  const { control } = useForm();

  return (
    <Input name="firstName" control={control} />
  )
}
```

**Comment by bluebill1049** [comment]
for react native, you will need to use useController or Controller.

take a look at this section: https://react-hook-form.com/get-started#ReactNative there is expo snack there as well.

**Reply by Noitidart** [reply]
Thank you sir, I udnerstand that, and used it as such in the code above, I think. Maybe I misunderstood. However I was wondering how react native avoids the re-rendering.

**Reply by bluebill1049** [reply]
oh we isolated the re-render without the Controller/useController so it doesn't impact your form/app.

**Reply by Noitidart** [reply]
Ohh! That's very interesting. I just verified that by putting a counter into the TextInput component. I will try creating a prototype of a direct manipulation based technique and ask you how it looks. Thanks for your kind and patience working with everyone's questions.

**Reply by bluebill1049** [reply]
Welcome.

**Reply by Noitidart** [reply]
I just wrote it so the textinput does not ever render either! :)

I had to make onChange and onBlur stable though, is this a sound assumption I made? That it's safe to make it stable -   https://github.com/react-hook-form/react-hook-form/pull/5113

Here I add a counter and we see that with `DirectHookedTextInput` it never re-renders! I did the stable trick from the PR above, is this trick safe? https://snack.expo.io/@noitidart/react-hook-form-v7---controller  - in this code here, the controller re-renders, but the render doesn't return anything new to commit. So it doesn't cause any react commit.

**Reply by Noitidart** [reply]
Here is a video, we see that with direct manipulation, the TextInput component never re-renders, the render counter stays at 1.

![](https://thumbs.gfycat.com/BronzeShamefulGosling-size_restricted.gif)

**Comment by Noitidart** [comment, marked as answer]
Answer to this question is in this comment - https://github.com/react-hook-form/react-hook-form/discussions/5111#discussioncomment-716613 thanks Bill!
